File: app/api/attribution.py
"""Transparent attribution ledger (Part 1, Feature 8).

GET  /attribution                        the ledger, newest outcome first
GET  /attribution/summary                which step and channel earned them
POST /attribution/recompute              run the sweep now (back-fills history)
GET  /leads/{lead_id}/attribution        what earned this prospect's outcomes

New paths. Nothing here writes an outcome; it only explains the ones that
already happened.
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.rate_limiting import enforce_rate_limit
from app.db.base import get_db
from app.db.models import User
from app.services import attribution, crm_service

router = APIRouter(tags=["attribution"])


@router.get("/attribution")
def list_attribution(
    outcome_kind: str | None = Query(default=None,
                                     pattern="^(meeting_booked|positive_reply|won)$"),
    strategy_id: uuid.UUID | None = None,
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """Every credited outcome with the touch behind it.

    Each row carries its `method` and `confidence`: a booking the prospect
    replied into is a fact, and "the last thing we sent before they booked" is
    a guess. Both are useful; presenting them identically is not."""
    return attribution.ledger(db, current_user.id, outcome_kind=outcome_kind,
                              strategy_id=strategy_id, limit=limit, offset=offset)


@router.get("/attribution/summary")
def attribution_summary(
    strategy_id: uuid.UUID | None = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """Which step and which channel actually earn outcomes.

    Every bucket carries a `certain` count beside its total, so an aggregate
    built mostly from last-touch guesses cannot be mistaken for one built from
    replies."""
    return attribution.summary(db, current_user.id, strategy_id=strategy_id)


@router.post("/attribution/recompute")
def recompute_attribution(
    limit: int = Query(default=500, ge=1, le=2000),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """Run the sweep now instead of waiting for the quarter-hourly one.

    Idempotent: UNIQUE (outcome_kind, outcome_id) means it can never
    double-credit, so this is safe to press repeatedly."""
    enforce_rate_limit(str(current_user.id), "attribution_recompute", "RATE_LIMIT_AI_ACTION")
    return attribution.run_sweep(db, limit=limit)


@router.get("/leads/{lead_id}/attribution")
def lead_attribution(lead_id: uuid.UUID, db: Session = Depends(get_db),
                     current_user: User = Depends(get_current_user)) -> list[dict]:
    """What earned this prospect's outcomes, with the message text itself."""
    lead = crm_service.owned_lead(db, lead_id, current_user)
    return attribution.for_lead(db, lead.id)

File: frontend/src/lib/api/attribution.ts
import { api } from "./client";
import type { AttributionEntry, AttributionSummary, OutcomeKind } from "../attribution";

/** Part 1 Feature 8 — the transparent attribution ledger. */

export interface AttributionPage {
  total: number;
  limit: number;
  offset: number;
  items: AttributionEntry[];
}

export function listAttribution(
  opts: { outcomeKind?: OutcomeKind; strategyId?: string; limit?: number } = {},
) {
  const params = new URLSearchParams();
  if (opts.outcomeKind) params.set("outcome_kind", opts.outcomeKind);
  if (opts.strategyId) params.set("strategy_id", opts.strategyId);
  params.set("limit", String(opts.limit ?? 100));
  return api<AttributionPage>(`/attribution?${params.toString()}`);
}

export const getAttributionSummary = (strategyId?: string) =>
  api<AttributionSummary>(
    `/attribution/summary${strategyId ? `?strategy_id=${strategyId}` : ""}`);

export const getLeadAttribution = (leadId: string) =>
  api<AttributionEntry[]>(`/leads/${leadId}/attribution`);

/** Idempotent server-side: UNIQUE(outcome_kind, outcome_id) means pressing
 *  this twice can never double-credit. */
export const recomputeAttribution = () =>
  api<{ checked: number; recorded: number; skipped: number; failed: number }>(
    "/attribution/recompute", { method: "POST" });
